### PR TITLE
HPCC-14689 Multi channel lookup/smart join implementation.

### PR DESCRIPTION
**Pull request by jakesmith**
Take advantage of a multi-channel slave setup, broadcasting to
each node vs slave and sharing the RHS table.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review.

NB: HPCC-14410 still needs to be done to take full advantage, but that is relatively small change and I have a separate branch for that.

File: system/jlib/jisem.hpp [hunk 1]
```diff
@@ -53,7 +53,10 @@ class jlib_decl InterruptableSemaphore : public Semaphore
             if (!_error)
                 _error = new InterruptedSemaphoreException;
             error.setown(_error);
-            signal();
+            if (1 == count)
```
Comment (ghalliday) [inline system/jlib/jisem.hpp]: trivial: Why the special case?

File: thorlcr/thorutil/thormisc.cpp [hunk 2]
```diff
@@ -816,9 +816,9 @@ void setClusterGroup(INode *_masterNode, IGroup *_rawGroup, unsigned slavesPerNo
     IArrayOf<INode> clusterGroupNodes, nodeGroupNodes;
     clusterGroupNodes.append(*LINK(masterNode));
     nodeGroupNodes.append(*LINK(masterNode));
-    for (unsigned p=0; p<slavesPerNode; p++)
+    for (unsigned s=0; s<channelsPerSlave; s++)
```
Comment (ghalliday) [inline thorlcr/thorutil/thormisc.cpp]: curious why the order has changed.

**Comment by ghalliday** [comment]
Comments on thlookupslave.cpp will be inline since not shown in diff.  Each one preceded by the line number.

102: Should allDoneWaiting be renamed now it is no longer a boolean.  what is it now counting?

**Comment by ghalliday** [comment]
97: Suggestion:  I assume slave means channel.  Is it worth consistently using node and channel rather than slave which is ambiguous?

**Comment by ghalliday** [comment]
260: Is this comment now out of date?

**Comment by ghalliday** [comment]
286: The removal of the -1 means it will now map to a different value.  I'm not sure that is intended.

**Comment by ghalliday** [comment]
1389: Is there a potential race condition?  setStopping doesn't have a cs.

**Comment by ghalliday** [comment]
1892: queryJobChannels() == 0?  Is this correct.  Could do with a comment why.

**Comment by ghalliday** [comment]
I'm struggling to review the logic.  I think it would be easier if the getRHS function was split up into smaller functions, which were named according to the processing stage.

**Comment by ghalliday** [comment]
The use tableProxy pointing at a shared entry looks suspicious for the many version - it contains state in the many variety - so it needs a separate instance per channel.

File: system/jlib/jisem.hpp (hunk 1 above)
Comment (jakesmith) [inline system/jlib/jisem.hpp]: Good point, I'll remove.

File: thorlcr/thorutil/thormisc.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/thorutil/thormisc.cpp]: To match the script slave/port assigned order.
init_thor* starts the slavePerNode processes with a port differential of : channelsperslave * localthorportinc.
e.g. with slavesPerNode = 2 and channelsPerSlave = 2, slave process 1 starts at 20100, slave process 2 = 20500.

Prior to this change, the order with the clusterGroup (when looking up ip:port of a channel slave) was wrong.

**Comment by jakesmith** [comment]
> 102: Should allDoneWaiting be renamed now it is no longer a boolean. what is it now counting?

Should change name, now it is not a boolean as you say.
It is count of channels waiting which will be signalled when a slaveStop call has been made for each slave.

**Comment by jakesmith** [comment]
> 97: Suggestion: I assume slave means channel. Is it worth consistently using node and channel rather than slave which is ambiguous?
>
>Or is channel 0..numChannelsPerSlave-1 and slave is 0..numSlaves-1? If so keep as it is.

The latter is correct.
For clarification:

There can be multiple channels per slave process.
There can be multiple slaves per node.

The cluster is formed of N slaves, which (assuming channelsPerslave is >1), will mean that there are some slaves on some channels and some on others, e.g. if channelsPerSlave=2
slave0 will be on channel 0
slave1 will be on channel 1
slave2 will be on channel 0
slave3 will be on channel 1
etc.

If slavesPerNode is >1, then the same is true, it's only the port ranges which change.

**Comment by jakesmith** [comment]
> 260: Is this comment now out of date?

Yes, will amend.

**Comment by jakesmith** [comment]
> 286: The removal of the -1 means it will now map to a different value. I'm not sure that is intended.
> Or is origin now 0 based instead of 1 based? It looks like that and myNode may have changed.

yes origin is now 0 based. The only thing that isn't is the MP send to #, which has to be 1 based for slaves, as the group contains the master @ 0

**Comment by jakesmith** [comment]
> 1389: Is there a potential race condition? setStopping doesn't have a cs.

Looks like you're right, and that I should add a CS.

**Comment by jakesmith** [comment]
> 1892: queryJobChannels() == 0? Is this correct. Could do with a comment why. Should it be checking channelNumber?

It is wrong, but was supposed to checking queryJobChannels().
However should be:
```
1 == queryJob().queryJobChannels()
```

It's checking to see if multiple channels, if not, it can use a regular row loader, if not it needs to be prepared for different spilling and rehashing to split source row arrays into separate channels on a spill event.

**Comment by jakesmith** [comment]
> I'm struggling to review the logic. I think it would be easier if the getRHS function was split up into smaller functions, which were named according to the processing stage.

I agree hard to follow, I'll try and introduce some granularity to it, with some new methods..

**Comment by jakesmith** [comment]
> The use tableProxy pointing at a shared entry looks suspicious for the many version - it contains state in the many variety - so it needs a separate instance per channel.

Hm, you're correct. Need separate HT, or at least separate HtEntry currentHashEntry

**Comment by ghalliday** [comment]
@jakesmith reassign back to you for the moment.

**Comment by jakesmith** [comment]
@ghalliday - please review, keeping refactor as separate commit for now.

**Comment by ghalliday** [comment]
One minor general comment.  Some classes e.g., CInMemJoinBase have large numbers of members.  If memory is a consideration then the order of the fields is significant - it would be better to place all the boolean fields together at the end, otherwise they can waste 7 bytes each (e.g., exclude).
In this case it doesn't matter much since there are only a few instances.

**Comment by jakesmith** [comment]
@ghalliday - noticed minor issue whilst looking at something else, pushed and squashed change.

**Comment by ghalliday** [comment]
thlookupjoinslave.cpp(1368)  broadcaster->cancel();
should this check if (broadcaster)

**Comment by ghalliday** [comment]
..slave.cpp(1890)   broadcaster->reset()
is a local smart join valid, and if so will this crash?

**Comment by ghalliday** [comment]
(2085) WHY??
Good question - possibly add a jira to discuss.

**Comment by ghalliday** [comment]
2126: Minor: "Failing over to *Standard* Join"

**Comment by ghalliday** [comment]
Looking pretty good after a first pass.  Of course we should also execute the hash lookups in parallel as part of the parallel activities work!

**Comment by jakesmith** [comment]
>>(2085) WHY??
>Good question - possibly add a jira to discuss.

Comment is stale, it does not failover to standard join if someone else does now.

**Comment by jakesmith** [comment]
>2126: Minor: "Failing over to Standard Join"

Will change.

**Comment by jakesmith** [comment]
>thlookupjoinslave.cpp(1368) broadcaster->cancel();
>should this check if (broadcaster)
>e.g., if local?
> ..slave.cpp(1890) broadcaster->reset()
>is a local smart join valid, and if so will this crash?

Well in fact a broadcaster instance is currently created regardless of whether local or not.
And those lines will not cause an issue, but are surplus to requirements in local mode.

**Comment by jakesmith** [comment]
@ghalliday - please see changes.

I did consider splitting classes down local vs global implementation, but they share a lot in common. It can wait for another PR if feeling bored.

**Comment by ghalliday** [comment]
@jakesmith please rebase, and let me know when you have resolved the minor issues you spotted.

**Comment by jakesmith** [comment]
@ghalliday - rebased and fixed some issues found whilst working on HPCC-14410

See 2nd commit.

**Comment by jakesmith** [comment]
NB: There is still the unexplained issue I talked to you about yesterday where compactRows hits an invalid pointer occasionally, giving an error like:
```
00000936 2016-02-10 15:15:17.069  9490 12601 "Release: Invalid pointer(base=0x7f03c9f400a0) 0x7f03c9f400a8 0"
```

It is always on the 2nd call (it does for each rowarray, 1 per channel)) and always traces *baseptr as 0.
Stack for reference:
```
000004F1 2016-02-10 16:51:28.134  9979 11061 "calling compactRows (numRows=3892988, rows=0x7f0589400080"
000004F2 2016-02-10 16:51:28.145  9979 11061 "Release: Invalid pointer(base=0x7f05815c00c0) 0x7f05815c00c8 0"
000004F3 2016-02-10 16:51:28.145  9979 11061 "Backtrace:"
000004F4 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libjlib.so(+0xe1628) [0x7f05a6efb628]"
000004F5 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libroxiemem.so(_ZN8roxiemem16FixedSizeHeaplet12noteReleasedEPKv+0x14b) [0x7f05a52ce6cb]"
000004F6 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libroxiemem.so(_ZN8roxiemem16FixedSizeHeaplet9doMoveRowEPcPKv+0x5f) [0x7f05a52cee0f]"
000004F7 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libroxiemem.so(_ZN8roxiemem12CChunkedHeap10compactRowEPKvRNS_16HeapCompactStateE+0x97) [0x7f05a52c4767]"
000004F8 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libroxiemem.so(_ZN8roxiemem19CChunkingRowManager11compactRowsEmPPKv+0x5a) [0x7f05a52cef7a]"
000004F9 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libgraph_lcr.so(_ZN22CThorExpandingRowArray7compactEv+0x72) [0x7f05a5da06e2]"
000004FA 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libgraph_lcr.so(_ZN22CThorSpillableRowArray7compactEv+0x46) [0x7f05a5da1c16]"
000004FB 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libactivityslaves_lcr.so(_ZN23CLookupJoinActivityBaseI13CLookupManyHTE20getGatheredRHSStreamEv+0x29e) [0x7f05a5876a4e]"
000004FC 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libactivityslaves_lcr.so(_ZN23CLookupJoinActivityBaseI13CLookupManyHTE24handleFailoverToLocalRHSER22CThorExpandingRowArrayP8ICompare+0x10e) [0x7f05a5876f1e]"
000004FD 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libactivityslaves_lcr.so(_ZN23CLookupJoinActivityBaseI13CLookupManyHTE6getRHSEb+0x697) [0x7f05a5877847]"
000004FE 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libactivityslaves_lcr.so(_ZN23CLookupJoinActivityBaseI13CLookupManyHTE7nextRowEv+0x7c) [0x7f05a5877d3c]"
000004FF 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libactivityslaves_lcr.so(_ZN22AggregateSlaveActivity7nextRowEv+0x96) [0x7f05a5807e56]"
00000500 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libactivityslaves_lcr.so(_ZN20CResultSlaveActivity7processEv+0x63) [0x7f05a58a1eb3]"
00000501 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libactivityslaves_lcr.so(_ZN20ProcessSlaveActivity4mainEv+0x6e) [0x7f05a57ff86e]"
00000502 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libjlib.so(_ZN19CThreadedPersistent4mainEv+0x4c) [0x7f05a6ee656c]"
00000503 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libjlib.so(_ZN19CThreadedPersistent8CAThread3runEv+0x10) [0x7f05a6fac8f0]"
00000504 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libjlib.so(_ZN6Thread5beginEv+0x2d) [0x7f05a6fa544d]"
00000505 2016-02-10 16:51:28.146  9979 11061 "  /opt/HPCCSystems/lib/libjlib.so(_ZN6Thread11_threadmainEPv+0x1e) [0x7f05a6fa6dde]"
00000506 2016-02-10 16:51:28.146  9979 11061 "  /lib/x86_64-linux-gnu/libpthread.so.0(+0x8182) [0x7f05a4b6b182]"
00000507 2016-02-10 16:51:28.146  9979 11061 "  /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f05a489847d]"

```

**Comment by jakesmith** [comment]
@ghalliday - managed to provoke the compactRows 'Invalid pointer' bug, I have opened a JIRA:
https://track.hpccsystems.com/browse/HPCC-14980

**Comment by ghalliday** [comment]
Merged - the invalid pointer issue is fixed in a separate PR
